### PR TITLE
rust: add `net::filter` module

### PR DESCRIPTION
**Pull request by wedsonaf**
This allows us to add network packet filters written in Rust.

Signed-off-by: Wedson Almeida Filho <wedsonaf@google.com>

### CODE REVIEW

File: rust/kernel/net/filter.rs
Comment (bjorn3) [inline rust/kernel/net/filter.rs]: If the filter gives the id of a non-existent user-space queue what will happen? Will it result in a memory safety issue, an oops or something else?

Comment (wedsonaf) [reply rust/kernel/net/filter.rs]: Good question: the packet is dropped. Looking at the code now, I see that there's a way to tell the caller to accept the packet if the queue doesn't exist. Let me add that piece here.

[The case when `NF_QUEUE` is returned is handled [here](https://elixir.bootlin.com/linux/v5.17/source/net/netfilter/core.c#L631); the decision on whether to accept or reject the packet happens in the `nf_queue` implementation [here](https://elixir.bootlin.com/linux/v5.17/source/net/netfilter/nf_queue.c#L244). `ESRCH` is returned if there are no [queue handlers](https://elixir.bootlin.com/linux/v5.17/source/net/netfilter/nf_queue.c#L169) (the infrastructure is pluggable, but only one [implementation](https://elixir.bootlin.com/linux/v5.17/source/net/netfilter/nfnetlink_queue.c#L1567) exists) or if a queue isn't [found](https://elixir.bootlin.com/linux/v5.17/source/net/netfilter/nfnetlink_queue.c#L788).]

Comment (wedsonaf) [reply rust/kernel/net/filter.rs]: I added this piece, PTAL.

**Review by bjorn3 — APPROVED** [review]
LGTM, but I think someone else with more knowledge about the behavior of the netfilter subsystem should take a look too.

**Comment by wedsonaf** [comment]
> LGTM, but I think someone else with more knowledge about the behavior of the netfilter subsystem should take a look too.

Thanks for reviewing, @bjorn3 -- I will ping the upstream maintainers to gauge their interest in reviewing and potentially eventually maintaining this.

**Pull request merged**
